source/agora/test/BanManager.d

```diff
@@ -79,42 +80,42 @@ unittest
     // leftover txs which node 3 will reject due to its filter
     Transaction[] left_txs;
 
-    auto new_tx = genBlockTransactions(4);
-    left_txs ~= new_tx;
-    new_tx.each!(tx => node_1.putTransaction(tx));
+    foreach (block_idx; 0 .. 4)
+    {
+        auto new_tx = genBlockTransactions(1);
+        left_txs ~= new_tx;
+        new_tx.each!(tx => node_1.putTransaction(tx));
+
+        [node_1, node_2].enumerate.each!((idx, node) =>
+            retryFor(node.getBlockHeight() == 1 + (block_idx + 1),
+                4.seconds,
+                format("Block %s Node %s has block height %s. Expected: %s",
+                    block_idx + 1, idx, node.getBlockHeight().to!string, 1 + (block_idx + 1))));
+
+        retryFor(node_3.getBlockHeight() == 1, 1.seconds, node_3.getBlockHeight().to!string);
+    }
 
     // wait for node 3 to be banned and all putTransaction requests to time-out
     Thread.sleep(2.seconds);
```

Comment:
Is this still relevant ?

Reply:
Yeah because it still takes actual real world time for a node to be banned.

I can see two ways of making this better:
- Use a virtual clock so we can "fast-forward" events. It's still tricky to write code for this. I've attempted this once.
- Add an event handling mechanism, so we can wait until some event fires. Something like this:

```D
class Events
{
    void onBannedNode (Node node) { ... }
    void onWhatever (...) { ... }
}
```

Then in the test-suite we could provide a subclass of `Events`, give that to the `Node`, and we could listen and wait for events. Instead of this `Thread.sleep(2.seconds);` call, we could just wait for the `onBannedNode` call.

Reply:
I wonder if anyone has implemented https://en.wikipedia.org/wiki/Server-sent_events in D